superset/connectors/sqla/models.py

```diff
@@ -804,10 +837,24 @@ def query(self, query_obj):
                 df = df.rename(index=str, columns=self.mutated_labels)
         except Exception as e:
             status = utils.QueryStatus.FAILED
-            logging.exception(e)
+            logging.exception(f'Query {sql} on schema {self.schema} failed')
             error_message = (
                 self.database.db_engine_spec.extract_error_message(e))
 
+        labels_expected = query_str_ext.labels_expected
+        if df is not None and \
+                not df.empty and \
+                not self.database.db_engine_spec.db_respects_given_column_labels and \
+                labels_expected is not None:
+            if len(df.columns) != len(labels_expected):
+                error_message = (f'For {sql}, df.columns: {df.columns}' +
+                                 f' differs from {labels_expected}')
+                status = utils.QueryStatus.FAILED
+            else:
+                logging.info(f'For query {sql}, renaming the df.columns from'
+                             f' {df.columns} to {labels_expected}')
+                df.columns = labels_expected
+
```

Comment:
This seems very specific to Pinot. I wonder if this could be dealt with in `db_engine_specs`, too by adding a method called `mutate_column_labels(...)` or similar?

Reply:
As I understand it, `labels_expected` is what we want the columns in the resulting dataframe to be called based on the original unmutated `select_exprs` above. In this sense I believe it is quite similar to what `mutate_label()` in db_engine_spec does, as it is something that we have to do due to quirkiness of the db. To illustrate with an extreme example, we might have to implement a db_engine_spec for an engine that decides to always return an extra trailing dummy column with null values, i.e. `len(df.columns) == len(labels_expected) + 1`. To cater for this possibility, it would be optimal for `sqla/models.py` to not have to be aware of this, but rather be able to push that handling to the respective db engine spec. For this scenario I think having a `def mutate_df_columns(df: pd.DataFrame, labels_expected: Sequence[str]) -> pd.DataFrame` in `db_engine_specs.py` would be quite appropriate; `BaseEngineSpec` would just return the unmutated df, while `PinotEngineSpec` would perform the above logic, and hypothetical `TrailingDummyColEngineSpec` would remove the redundant column. Etc etc.

P.S. The rename logic I have done is rather controversial and does bend the rules, and could probably have been solved more elegantly, but would have required a bigger refactor that would have been unfeasible. But as it was a common problem for several dbs, the current solution was deemed a good compromise.

Reply:
I think that the logic belongs in the DB API driver instead of Superset — it would also benefit other tools that use the driver.

I wrote a Google Spreadsheets connector that has the same problem, since the SQL API supported is very crude and doesn't accept aliases. Maybe you can reuse the helper function I created? It's based on `moz-sql-parser` instead of `sqlparse`:

https://github.com/betodealmeida/gsheets-db-api/blob/master/gsheetsdb/query.py#L93

Reply:
I am wondering if I could do that as a part of the future work :). 

I agree with you that the ideal way to do this is to have the DBAPI deal with this and that requires doing a re-parse of the query to infer the required label names. Given all the difficulty of parsing (or unparsing) PQL, I would rather do this more properly than hack something in here to unblock this diff. 

I anyway have to rework the pinot-dbapi to integrate it with dbapihelper, so I can also consider integrating it with this sql parser then too. 

I am just anxious of hanging onto this PR in our internal branch here that increases our merge burden.

Thanks